---
layout: post
title: RNA-seq
subtitle: RNA-seq 
gh-repo: leequn/leequn.github.io
gh-badge: [star, fork, follow]
tags: [Bioinformatics]
comments: true
---

RNA-seq

## sra to fastq
~~~shell
for i in *.sra
do
echo ${i:0:9}
fastq-dump --split-3 $i
done
~~~

## trim fastq
**single**
~~~shell
for i in *.fastq
do 
trim_galore -q 30 -stringency 3 -length 20 --phred33 -clip_R1 3 -e 0.1 -dont_gzip -o trim  $i
done
~~~

**paired**
~~~shell
for i in *_1.fq.gz
do
echo ${i:0:9}
trim_galore -q 30 -stringency 3 -length 20 --phred33 -fastqc -clip_R1 3 -clip_R2 3 -e 0.1 -dont_gzip -paired ${i:0:9}_1.fq ${i:0:9}_2.fq
done
~~~

## map
**single**
~~~shell
for i in *.fq
do
echo ${i:0:10}
hisat2 -p 8 --dta -x index/mus -U $i -S sam/${i:0:10}.sam
done
~~~

**paired**
~~~shell
for i in *_1_val_1.fq
do
echo ${i:0:2}
hisat2 -p 8 --dta -x suoyin/pig_tran -1 $i -2 ${i:0:2}_2_val_2.fq -S ${i:0:2}.sam
done
~~~

## sam to bam
~~~shell
for i in *.sam
do 
echo ${i:0:10}
samtools sort -@ 8 -o ${i:0:10}.bam ${i:0:10}.sam
done
~~~
## quantification
~~~shell
for i in *.bam
do
echo ${i:0:10}
stringtie -e -B -p 8 -G Mus_musculus.GRCm38.90.gtf -o ${i:0:10}/${i:0:10}_mus.gtf  $i
done
~~~
## exp
**gene exp**
~~~R
rm(list=ls())
library(ballgown)
filenames <- list.files('ballgown')
dir <- paste('ballgown/',filenames,sep='')
n <- length(filenames)
for (i in 1:n){
	bg <- ballgown(samples=dir[i],meas='all')
	gene_expression <- gexpr(bg)
	path <- paste("exp/",filenames[i],".csv",sep="")
	write.csv(gene_expression,file=path)
}
~~~

**diff exp**
~~~R
library(ballgown)
library(genefilter)
library(dplyr)
library(devtools)
phenotype = read.csv("phenotype.csv",header=T,sep="\t")
head(phenotype)
#--------------------
#         ids    sex    population
#1 SRR6466413 female A
#2 SRR6466414 female A
#3 SRR6466415 female B
#4 SRR6466417 male A
#5 SRR6466418 male B
#6 SRR6466419 male B
#--------------------
bg_data = ballgown(dataDir="ballgown",samplePattern="SRR",pData=phenotype)
bg_filt = subset(bg_data,"rowVars(texpr(bg_data)) >1",genomesubset=TRUE)
results_transcripts = stattest(bg_filt, feature="transcript",covariate="sex", adjustvars= c("population"), getFC=TRUE, meas="FPKM")
results_genes = stattest(bg_filt, feature="gene",covariate="sex", adjustvars =c("population"), getFC=TRUE,meas="FPKM")
results_transcripts = data.frame(geneNames=ballgown::geneNames(bg_filt),geneIDs=ballgown::geneIDs(bg_filt), results_transcripts)
results_transcripts = arrange(results_transcripts,pval)
results_genes = arrange(results_genes,pval)
write.csv(results_transcripts, "transcript_results.csv", row.names=FALSE)
write.csv(results_genes, "gene_results.csv", row.names=FALSE)
~~~


