---
layout: post
title: Single cell
subtitle: Single-cell analysis
gh-repo: leequn/leequn.github.io
gh-badge: [star, fork, follow]
tags: [Bioinformatics, Single Cell]
comments: true
---

# Single-cell analysis

## Introduction
### Why scRNA-seq
- 1. explore which cell types are present in a tissue
- 2. identify unknown/rare cell types or states
- 3. elucidate the changes in gene expression during differentiation processes or across time or states
- 4. identify genes that are differentially expressed in particular cell types between conditions (e.g. treatment or disease)
- 5. explore changes in expression among a cell type while incorporating spatial, regulatory, and/or protein information

### Challenges of scRNA-seq analysis
- 1. Large volume of data
- 2. Low depth of sequencing per cell
- 3. Biological variability across cells/samples
	- Transcriptional bursting
	- Varying rates of RNA processing
	- Continuous or discrete cell identities (e.g. the pro-inflammatory potential of each individual T cell)
	- Environmental stimuli
	- Temporal changes
- 4. Technical variability across cells/samples
	- Cell-specific capture efficiency
	- Library quality
	- Amplification bias
	- Batch effects

## 

## References

{: .box-note}
**Note:** 1. [Introduction to Single-cell RNA-seq](https://hbctraining.github.io/scRNA-seq/schedule/)
